src/critnib/critnib.c

```diff
@@ -176,37 +182,64 @@ static inline unsigned slice_index(word key, sh_t shift) {
  * critnib_new -- allocates a new critnib structure
  */
 struct critnib *critnib_new(void) {
-    struct critnib *c = Zalloc(sizeof(struct critnib));
-    if (!c) {
+    umf_ba_linear_pool_t *pool_linear =
+        umf_ba_linear_create(0 /* minimal pool size */);
+    if (!pool_linear) {
         return NULL;
     }
 
-    c->mutex = util_mutex_create();
+    struct critnib *c =
+        umf_ba_linear_alloc(pool_linear, sizeof(struct critnib));
+    if (!c) {
+        goto err_destroy_pool_linear;
+    }
+
+    c->pool_linear = pool_linear;
+
+    c->mutex = util_mutex_init(
+        umf_ba_linear_alloc(pool_linear, util_mutex_get_size()));
     if (!c->mutex) {
-        free(c);
-        return NULL;
+        goto err_destroy_pool_linear;
+    }
+
+    c->pool_nodes = umf_ba_create(sizeof(struct critnib_node));
+    if (!c->pool_nodes) {
+        goto err_util_mutex_destroy;
+    }
+
+    c->pool_leaves = umf_ba_create(sizeof(struct critnib_leaf));
+    if (!c->pool_leaves) {
+        goto err_destroy_pool_nodes;
     }
 
     VALGRIND_HG_DRD_DISABLE_CHECKING(&c->root, sizeof(c->root));
     VALGRIND_HG_DRD_DISABLE_CHECKING(&c->remove_count, sizeof(c->remove_count));
 
     return c;
+
+err_destroy_pool_nodes:
+    umf_ba_destroy(c->pool_nodes);
+err_util_mutex_destroy:
+    util_mutex_destroy_not_free(c->mutex);
+err_destroy_pool_linear:
+    umf_ba_linear_destroy(pool_linear); // free all its allocations and destroy
+    return NULL;
 }
 
 /*
  * internal: delete_node -- recursively free (to malloc) a subtree
  */
-static void delete_node(struct critnib_node *__restrict n) {
+static void delete_node(struct critnib *c, struct critnib_node *__restrict n) {
     if (is_leaf(n)) {
-        Free(to_leaf(n));
+        umf_ba_free(c->pool_leaves, to_leaf(n));
```

Comment:
I think we could remove also Malloc, Free and Zalloc symbols from code

Reply:
None of these functions is used in the critnib code.

Reply:
Then we can also remove them from utils_common.h I think only critnib used them.

Reply:
Removed